core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -367,46 +290,46 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
     debug("Controller resigning, broker id %d".format(config.brokerId))
     // de-register listeners
     deregisterIsrChangeNotificationListener()
-    deregisterReassignedPartitionsListener()
+    deregisterPartitionReassignmentListener()
     deregisterPreferredReplicaElectionListener()
 
     // shutdown delete topic manager
-    if (deleteTopicManager != null)
-      deleteTopicManager.shutdown()
+    if (topicDeletionManager != null)
+      topicDeletionManager.shutdown()
 
     // shutdown leader rebalance scheduler
-    if (config.autoLeaderRebalanceEnable)
-      autoRebalanceScheduler.shutdown()
-
-    inLock(controllerContext.controllerLock) {
-      // de-register partition ISR listener for on-going partition reassignment task
-      deregisterReassignedPartitionsIsrChangeListeners()
-      // shutdown partition state machine
-      partitionStateMachine.shutdown()
-      // shutdown replica state machine
-      replicaStateMachine.shutdown()
-      // shutdown controller channel manager
-      if(controllerContext.controllerChannelManager != null) {
-        controllerContext.controllerChannelManager.shutdown()
-        controllerContext.controllerChannelManager = null
-      }
-      // reset controller context
-      controllerContext.epoch=0
-      controllerContext.epochZkVersion=0
-      brokerState.newState(RunningAsBroker)
-
-      info("Broker %d resigned as the controller".format(config.brokerId))
+    kafkaScheduler.shutdown()
+    offlinePartitionCount.set(0)
+    preferredReplicaImbalanceCount.set(0)
+
+    // de-register partition ISR listener for on-going partition reassignment task
+    deregisterPartitionReassignmentIsrChangeListeners()
+    // shutdown partition state machine
+    partitionStateMachine.shutdown()
+    deregisterTopicChangeListener()
+    partitionModificationsListeners.keys.foreach(deregisterPartitionModificationsListener)
```

Comment:
We need to maintain some sort of mapping from path to PartitionModificationsListener objects somewhere, or at the very least some way of associating an existing IZkDataListener instance to its corresponding path because ZkClient's unsubscribe variants take both the path and corresponding listener instance. For example:
```
public void unsubscribeDataChanges(String path, IZkDataListener dataListener)
```
I've considered defining wrapper listener classes comprised of a zkclient, raw zkclient listener, and path such that you can just call MyWrapperListener.subscribe()/unsubscribe() and it'll internally call zkClient.subscribeDataChanges(path, rawZkClientListener) but decided to leave it out of this patch to minimize the changes.